session/xmlrpc.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/softlayer/softlayer-go/sl"
+)
+
+// Used to pool the clients created per service
+// so as to re-use service clients created previously
+var xmlRpcClients = map[string]*xmlrpc.Client{}
+
+// Debugging RoundTripper
+type debugRoundTripper struct{}
+
+func (mrt debugRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	log.Println("->>>Request:")
+	dumpedReq, _ := httputil.DumpRequestOut(request, true)
+	log.Println(string(dumpedReq))
+
+	response, err := http.DefaultTransport.RoundTrip(request)
+	if err != nil {
+		log.Println("Error:", err)
+		return response, err
+	}
+
+	log.Println("\n\n<<<-Response:")
+	dumpedResp, _ := httputil.DumpResponse(response, true)
+	log.Println(string(dumpedResp))
+
+	return response, err
+}
+
+// XML-RPC Transport
+type XmlRpcTransport struct {
+	Timeout time.Duration
+}
+
+const DefaultXmlRpcTimeout = time.Second * 30
+
+// err = r.Session.DoRequest("SoftLayer_Account", "activatePartner", params, &r.Options, &resp)
+func (x *XmlRpcTransport) DoRequest(
+	sess *Session,
+	service string,
+	method string,
+	args []interface{},
+	options *sl.Options,
+	pResult interface{},
+) error {
+
+	client, ok := xmlRpcClients[service]
+	if !ok {
+		var roundTripper http.RoundTripper
+		var err error
+
+		if sess.Debug {
+			roundTripper = debugRoundTripper{}
+		}
+
+		timeout := DefaultXmlRpcTimeout
+		if x.Timeout != 0 {
+			timeout = x.Timeout
+		}
+
+		client, err = xmlrpc.NewClient(
+			fmt.Sprintf("%s/%s", sess.Endpoint, service),
+			roundTripper,
+			timeout,
+		)
+		if err != nil {
+			return fmt.Errorf("Could not create an xmlrpc client for %s: %s", service, err)
+		}
+
+		xmlRpcClients[service] = client
+	}
+
+	authenticate := map[string]interface{}{}
+	if sess.UserName != "" {
+		authenticate["username"] = sess.UserName
+	}
+
+	if sess.APIKey != "" {
+		authenticate["apiKey"] = sess.APIKey
+	}
+
+	if sess.UserId != 0 {
+		authenticate["userId"] = sess.UserId
+		authenticate["complexType"] = "PortalLoginToken"
+	}
+
+	if sess.AuthToken != "" {
+		authenticate["authToken"] = sess.AuthToken
+		authenticate["complexType"] = "PortalLoginToken"
+	}
+
+	headers := map[string]interface{}{}
+	if len(authenticate) > 0 {
+		headers["authenticate"] = authenticate
+	}
+
+	if options.Id != nil {
+		headers[fmt.Sprintf("%sInitParameters", service)] = map[string]int{
+			"id": *options.Id,
+		}
+	}
+
+	mask := options.Mask
+	if mask != "" {
+		if !strings.HasPrefix(mask, "mask[") {
+			mask = fmt.Sprintf("mask[%s]", mask)
+		}
+		headers["SoftLayer_ObjectMask"] = map[string]string{"mask": mask}
+	}
+
+	if options.Filter != "" {
+		// FIXME: This json unmarshaling presents a performance problem,
+		// since the filter builder marshals a data structure to json.
+		// This then undoes that step to pass it to the xmlrpc request.
+		// It would be better to get the umarshaled data structure
+		// from the filter builder, but that will require changes to the
+		// public API in Options.
```

Comment:
Hmm.  Are you suggesting that Options.Filter should be `interface{}` so that it can store either a string (for specifying a raw filter) or a `Filters` object, and that `Build()` should really be left to the transport?